wasm/src/main/java/com/dylibso/chicory/wasm/io/BufferWasmInputStream.java

```diff
@@ -0,0 +1,88 @@
+package com.dylibso.chicory.wasm.io;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A WASM input stream backed by a byte buffer.
+ */
+class BufferWasmInputStream extends WasmInputStream {
+    private final ByteBuffer buffer;
+
+    BufferWasmInputStream(final ByteBuffer buffer) {
+        super();
+        // make our position independent, and zero-based
+        this.buffer = buffer.slice();
```

Comment:
Here we are creating a new ByteBuffer: https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#slice--
getting independent on the index, but still sharing the content.
I'm not sure about this decision:
- if we wanna be independent of the provided `ByteBuffer` we should `duplicate` instead
- If we wanna share the original `ByteBuffer` not considering the current index is not the behavior I would expect as a user

Reply:
> We use a slice because we need the byte content to extend from position zero

I'm failing to get the "why", I think that there might be use-cases where preserving the current position would be the expected behavior.

Reply:
The contract of this class is that reading extends from the start to the end (no random access is allowed), and that the `position` of the stream reports the number of bytes consumed by the stream. So, we have to ensure that the buffer position increases monotonically (hence having our own copy); after that, using the buffer position as the stream position follows logically. But in order to do so, the position must start at zero, hence the slice.

Reply:
> The contract of this class is that reading extends from the start to the end

Do you have a reference? I'm failing to find this detail other than in the implementation.

I see a possible use-case where you are reading something that "contains" WASM.
In this case, the user might want to consume part of the Stream, and directly pass the buffer to be consumed starting from the current position.
This doesn't break monotonicity but doesn't need to start at 0.

Reply:
Implicit in the definition of "stream" is that it progresses monotonically from start to end. I'm not aware of any "stream" type class in any commonly used Java API that does not behave this way.

The use case you illustrate is supported directly by this API. If the user passes the buffer to be consumed, the stream will read from that buffer starting at the buffer's position and extending to the buffer's limit. The stream's reported position will start at zero. The user's buffer position and limit is not affected, and the user can independently change their buffer position and limit without affecting the stream position.